### PR TITLE
Transforms v2 Compliance

### PR DESCRIPTION
**Pull request by broomhead**
Issue #170 inspired some work to update the transforms in this project to be transforms v2 compliant. Some of the code was still doing transforms the v1 way while being shoved into the v2 pipeline and throwing errors and causing problems.

I propose making a clean break to v2 transforms as well as v2 tv_tensor types. To this end, I have removed _misc.py completely as its attempts at backwards compatibility aren't going to work moving forward. _misc.py tried to support back to 0.15.2, but Torch Vision is at 0.22 today. There's enough time that new builds should be pulling in a version more recent anyway.

Only a couple files have significant changes; the rest are 1-2 line changes to imports to switch to functional and transforms v2, which are drop-in ready. You'll see changes like

```
import torchvision.transforms as T
import torchvision.transforms.functional as F
```
to
```
import torchvision.transforms.v2 as T
import torchvision.transforms.functional.v2 as F
```

I'll walk through the meatier parts.

### CODE REVIEW

File: configs/dfine/include/dataloader.yml [hunk 1]
```diff
@@ -5,13 +5,16 @@ train_dataloader:
       ops:
         - {type: RandomPhotometricDistort, p: 0.5}
         - {type: RandomZoomOut, fill: 0}
-        - {type: RandomIoUCrop, p: 0.8}
-        - {type: SanitizeBoundingBoxes, min_size: 1}
+        - type: RandomApply
+          p: 0.8
+          transforms: 
+            - type: RandomIoUCrop
```
Comment (broomhead) [inline configs/dfine/include/dataloader.yml:11]: RandomApply exists as a V2 transform, but it doesn't work with the yaml config parser. So I made a thin version of RandomApply that allows us to set a p value for when a list of transforms should fire. _transforms.py will clarify.

File: configs/dfine/include/dataloader.yml [hunk 2]
```diff
@@ -5,13 +5,16 @@ train_dataloader:
...
         - {type: RandomHorizontalFlip}
-        - {type: Resize, size: [640, 640], }
+        - {type: Resize, size: [640, 640] }
         - {type: SanitizeBoundingBoxes, min_size: 1}
         - {type: ConvertPILImage, dtype: 'float32', scale: True}
-        - {type: ConvertBoxes, fmt: 'cxcywh', normalize: True}
+        - {type: ConvertBoundingBoxFormat, format: 'CXCYWH'}
```
Comment (broomhead) [inline configs/dfine/include/dataloader.yml:16]: This is the right way to transform BoundingBoxes with the library-provided ConvertBoundingBoxFormat. Since the code now guarantees we have TV Tensor types, this works great.

File: configs/dfine/include/dataloader.yml [hunk 3]
```diff
@@ -5,13 +5,16 @@ train_dataloader:
...
+        - {type: NormalizeAnnotations}
```
Comment (broomhead) [inline configs/dfine/include/dataloader.yml:17]: I felt it cleaner to separate converting types from normalizing them. Previously, `normalize=True` was an option to the prior bounding box conversion transform.

File: src/core/workspace.py [hunk 4]
```diff
@@ -169,10 +169,6 @@ def create(type_or_name, global_cfg=GLOBAL_CONFIG, **kwargs):
...
     # TODO hard code
     module_kwargs = {k: v for k, v in module_kwargs.items() if not k.startswith("_")}
-
-    # TODO for **kwargs
-    # extra_args = set(module_kwargs.keys()) - set(arg_names)
-    # if len(extra_args) > 0:
-    #     raise RuntimeError(f'Error: unknown args {extra_args} for {module}')
-
+    # allow extra kwargs passed to create to override or add
+    module_kwargs.update(kwargs)
```
Comment (broomhead) [inline src/core/workspace.py:173]: I seem to recall I needed to add this for some config parts to parse; regardless, it makes the extra kwargs work and doesn't break existing functionality.

File: src/data/transforms/_transforms.py [hunk 5]
```diff
@@ -13,30 +13,20 @@
...
+from ...core import register, create
 
-torchvision.disable_beta_transforms_warning()
 
+torchvision.disable_beta_transforms_warning()
 
+ConvertBoundingBoxFormat = register()(T.ConvertBoundingBoxFormat)
+RandomCrop = register()(T.RandomCrop)
+RandomHorizontalFlip = register()(T.RandomHorizontalFlip)
+RandomIoUCrop = register()(T.RandomIoUCrop)
```
Comment (broomhead) [inline src/data/transforms/_transforms.py:25]: In V2 transforms, some of these have an optional p parameter so we don't need to implement local copies anymore. If there's a transform without a `p` param, use the supplied `RandomApply`

File: src/data/transforms/_transforms.py [hunk 6]
```diff
@@ -46,94 +36,95 @@ def __init__(
...
+            t_type = kwargs.pop("type")
+            inner_transform = create(t_type, **kwargs)
+            instances.append(inner_transform)
+        
+        super().__init__(torch.nn.ModuleList(instances), p=p)
 
 
 @register()
-class PadToSize(T.Pad):
+class Letterbox(T.Transform):
```
Comment (broomhead) [inline src/data/transforms/_transforms.py:65]: I threw in a transform I made for my purposes. It essentially does exactly what `PadToSize` was doing, but instead of padding the image from the top and left, it centers it in "letterbox" style. This allows the image to be 640x640 without skewing the image ratio. The configs don't seem to be using it but I included it as a bonus.

File: src/data/transforms/_transforms.py [hunk 7]
```diff
@@ -46,94 +36,95 @@ def __init__(
...
-
-    def __call__(self, *inputs: Any) -> Any:
-        outputs = super().forward(*inputs)
-        if len(outputs) > 1 and isinstance(outputs[1], dict):
-            outputs[1]["padding"] = torch.tensor(self.padding)
-        return outputs
-
-
-@register()
-class RandomIoUCrop(T.RandomIoUCrop):
```
Comment (broomhead) [inline src/data/transforms/_transforms.py:89]: Transforms v2 supplies this exact transform, with an optional p param. No need to rewrite here.

File: src/data/dataset/coco_dataset.py [hunk 8]
```diff
@@ -60,14 +61,14 @@ def load_item(self, idx):
...
 
         target["idx"] = torch.tensor([idx])
+        canvas_size = F.get_size(image)
 
         if "boxes" in target:
-            target["boxes"] = convert_to_tv_tensor(
-                target["boxes"], key="boxes", spatial_size=image.size[::-1]
-            )
+            # COCO annotations are in XYWH format, but self.prepare() converted them to XYXY.
+            target["boxes"] = BoundingBoxes(target["boxes"], format="XYXY", canvas_size=canvas_size)
```
Comment (broomhead) [inline src/data/dataset/coco_dataset.py:68]: Just caught this bug. I had put XYWH, which is the correct COCO format, but the call above to prepare() was outputing the tensors in XYXY. All good now and the rendered samples look correct!
